### PR TITLE
Updates to Popup Focus

### PR DESCRIPTION
**Pull request by pmkhougaz**
### Change Log
- Adds focus mixin to enable use of `this.handleFocus` or `this.handleBlur`
- Correct focus for tetheredPopup components

EDIT: please take note of @awseward's comment below. the only relevant changes are files that are under the `src` directory. all else are built into the lib

### CODE REVIEW

**Comment by awseward** [comment]
@piercefrawley: Presumably we really only need to concern ourselves with changes you've made in the `src/` directory, correct?

File: src/List.jsx [hunk 1]
```diff
@@ -108,6 +108,7 @@ export default React.createClass({
 
     return (
       <ul
+        ref="foo"
```
Comment (awseward) [inline src/List.jsx]: Could we maybe come up with a more descriptive string here?

File: src/TetheredPopup.jsx [hunk 2]
```diff
@@ -103,12 +100,13 @@ module.exports = React.createClass({
 
     const placeholderEl = compat.findDOMNode(placeholder);
 
-    if (!placeholderEl) return null;
+    // if (!placeholderEl) return null;
```
Comment (awseward) [inline src/TetheredPopup.jsx]: Can we just remove this line?

File: src/mixins/FocusMixin.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      return
+
+    inst.setTimeout('focus', () => {
+      compat.batchedUpdates(() => {
+        if (didHandle) didHandle.call(inst, focused, event)
+
+        if (focused !== inst.state.focused) {
+          notify(handler, event)
+          if (inst.isMounted())
+            inst.setState({ focused })
```
Comment (awseward) [inline src/mixins/FocusMixin.js:22]: I understand this file may have been code pulled from upstream, but these unbracketed `if`s really make me nervous...

File: src/TetheredPopup.jsx (hunk 2 above)
Comment (pmkhougaz) [inline src/TetheredPopup.jsx]: yes to this and your first comment. I forgot to commit those changes

File: src/TetheredPopup.jsx [hunk 4]
```diff
@@ -194,7 +190,7 @@ module.exports = React.createClass({
     let focusComponent = content;
     let focusEl;
 
-    if(isFunction(getTetherFocus)) focusComponent = getTetherFocus();
+    focusComponent = getTetherFocus();
```
Comment (ksespinola) [inline src/TetheredPopup.jsx]: Should this be `const focusComponent = getTetherFocus()`?

File: src/mixins/FocusMixin.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    handleBlur(event) {
+      handleFocus(this, false, event)
+    },
+
+    @widgetEnabled
+    handleFocus(event) {
+      handleFocus(this, true, event)
+    }
+  }
+}
```
Comment (ksespinola) [inline src/mixins/FocusMixin.js]: Can you give some background into this? Something the author added to handle the use case we were missing?

**Comment by kespinola** [comment]
+1 pending comments

**Comment by awseward** [comment]
+1

**Comment by kespinola** [comment]
+1

**Pull request merged**